src/inmanta/plugins.py

```diff
@@ -259,188 +391,268 @@ def __init__(self, namespace: Namespace) -> None:
+        if self.var_kwargs is not None:
+            self.var_kwargs.resolve_type(self, self.resolver)
+
+        self.return_type.resolve_type(self, self.resolver)
+
+    def _load_signature(self, function: Callable[..., object]) -> None:
         """
-        Load the signature from the given python function
+        Load the signature from the given python function, and update the relevant attributes
+        of this object.
         """
-        arg_spec = inspect.getfullargspec(function)
+        # Reset relevant object attributes
+        self.args = list()
+        self.var_args = None
+        self.kwargs = dict()
+        self.var_kwargs = None
 
-        # Calculate at which index the default values start (for the non-keyword-only attributes)
-        default_start_for_args: Optional[int]
-        if arg_spec.defaults is not None:
-            default_start_for_args = len(arg_spec.args) - len(arg_spec.defaults)
-        else:
-            default_start_for_args = None
+        # Inspect the function to get its arguments and annotations
+        arg_spec = inspect.getfullargspec(function)
 
-        arguments: List[PluginArgument] = []
+        def get_annotation(arg: str) -> object:
+            """
+            Get the annotation for a specific argument, and if none exists, raise an exception
+            """
+            if arg not in arg_spec.annotations:
+                raise CompilerException(
+                    f"All arguments of plugin {repr(self.get_full_name())} should be annotated: "
+                    f"{repr(arg)} has no annotation"
+                )
 
-        def process_kw_only_args(argument_name: str, annotation: object) -> PluginArgument:
-            if arg_spec.kwonlydefaults and argument_name in arg_spec.kwonlydefaults:
-                default_value = arg_spec.kwonlydefaults[argument_name]
-                return PluginArgument(argument_name, annotation, is_kw_only_argument=True, default_value=default_value)
-            else:
-                return PluginArgument(argument_name, annotation, is_kw_only_argument=True)
+            return arg_spec.annotations[arg]
 
-        def process_regular_args(index: int, argument_name: str, annotation: object) -> PluginArgument:
-            if default_start_for_args is not None and default_start_for_args <= index:
-                default_value = arg_spec.defaults[default_start_for_args - index]
-                return PluginArgument(argument_name, annotation, is_kw_only_argument=False, default_value=default_value)
-            else:
-                return PluginArgument(argument_name, annotation, is_kw_only_argument=False)
+        # Make sure we have a return annotation even for implicit "null"
+        # If no return annotation (or a Falsy value) is provided, we write
+        # "null" in the annotations dict so that we can use the get_annotation
+        # helper defined above.
+        arg_spec.annotations["return"] = arg_spec.annotations.get("return") or "null"
```

Comment:
@sanderr @wouterdb 
I think this is not right, the previous implementation defaulted to `any` when not return annotation was provided (implicitly, it would actually simply not do any validation), this one defaults to none, and will validate that the returned value is None.  It is more inline with python, but also a breaking change, should I keep it consistent with the previous behavior?

Reply:
>  It is more inline with python

I don't think this is true. The following gives a mypy error for passing a string to `x`, but not for `return 1`.
```
def f(x: int):
    return 1


f("test")
```

So I think that indeed we should default to `any`.

Reply:
Oh, something I have to unlearn then.  So there is no discussion, I will fix it

Reply:
Fixed, and I added a test case for it